File: app/src/main/java/com/buffalo/controlefinancas/util/activity/ActivityUtil.java
package com.buffalo.controlefinancas.util.activity;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;

import java.io.Serializable;

public class ActivityUtil {

    private Context mContextActivity;
    private Class mClazz;
    private Bundle mBundle;
    private boolean mOnActivityResult;
    private int mRequestCode;

    public static class Builder {

        private Activity mContextActivity;
        private Class mClazz;
        private Bundle mBundle;
        private boolean mOnActivityResult;
        private int mRequestCode;

        private void initBuilder(Class aClazz) {
            this.mBundle = new Bundle();
            this.mClazz = aClazz;
        }

        public Builder(Activity aActivity, Class aClazz) {
            this.mContextActivity = aActivity;
            initBuilder(aClazz);
        }

        public Builder putBundleInt(String aKey, int aValue) {
            mBundle.putInt(aKey, aValue);
            return this;
        }

        public Builder putBundleLong(String aKey, long aValue) {
            mBundle.putLong(aKey, aValue);
            return this;
        }

        public Builder putBundleString(String aKey, String aValue) {
            mBundle.putString(aKey, aValue);
            return this;
        }

        public Builder putBundleSerializable(String aKey, Serializable aValue) {
            mBundle.putSerializable(aKey, aValue);
            return this;
        }

        public Builder putBundleDouble(String aKey, double aValue) {
            mBundle.putDouble(aKey, aValue);
            return this;
        }

        public Builder putBundleString(String aKey, float aValue) {
            mBundle.putFloat(aKey, aValue);
            return this;
        }

        public Builder onActivityResult(int aRequestCode) {
            mOnActivityResult = true;
            mRequestCode = aRequestCode;
            return this;
        }

        public ActivityUtil build() {
            return new ActivityUtil(this);
        }
    }

    private ActivityUtil(Builder aBuilder) {
        this.mContextActivity = aBuilder.mContextActivity;
        this.mBundle = aBuilder.mBundle;
        this.mClazz = aBuilder.mClazz;
        this.mRequestCode = aBuilder.mRequestCode;
        this.mOnActivityResult = aBuilder.mOnActivityResult;
        show();
    }

    private void show() {
        Intent lIntent = new Intent(mContextActivity, mClazz);
        lIntent.putExtras(mBundle);
        if(mOnActivityResult)
            if(mContextActivity instanceof Activity)
                ((Activity) mContextActivity).startActivityForResult(lIntent, mRequestCode);
            else
                mContextActivity.startActivity(lIntent);
        else
            mContextActivity.startActivity(lIntent);
    }
}

File: app/src/main/java/com/buffalo/controlefinancas/util/activity/ActivityUtil.kt
package com.buffalo.controlefinancas.util.activity

import android.os.Bundle
import android.app.Activity
import android.content.Context
import android.content.Intent
import java.io.Serializable

class ActivityUtil private constructor(aBuilder: Builder) {
    private val mContextActivity: Context
    private val mClazz: Class<*>?
    private val mBundle: Bundle?
    private val mOnActivityResult: Boolean
    private val mRequestCode: Int

    class Builder(val mContextActivity: Activity, aClazz: Class<*>) {
        var mClazz: Class<*>? = null
        var mBundle: Bundle? = null
        var mOnActivityResult = false
        var mRequestCode = 0
        private fun initBuilder(aClazz: Class<*>) {
            mBundle = Bundle()
            mClazz = aClazz
        }

        fun putBundleInt(aKey: String?, aValue: Int): Builder {
            mBundle!!.putInt(aKey, aValue)
            return this
        }

        fun putBundleLong(aKey: String?, aValue: Long): Builder {
            mBundle!!.putLong(aKey, aValue)
            return this
        }

        fun putBundleString(aKey: String?, aValue: String?): Builder {
            mBundle!!.putString(aKey, aValue)
            return this
        }

        fun putBundleSerializable(aKey: String?, aValue: Serializable?): Builder {
            mBundle!!.putSerializable(aKey, aValue)
            return this
        }

        fun putBundleDouble(aKey: String?, aValue: Double): Builder {
            mBundle!!.putDouble(aKey, aValue)
            return this
        }

        fun putBundleString(aKey: String?, aValue: Float): Builder {
            mBundle!!.putFloat(aKey, aValue)
            return this
        }

        fun onActivityResult(aRequestCode: Int): Builder {
            mOnActivityResult = true
            mRequestCode = aRequestCode
            return this
        }

        fun build(): ActivityUtil {
            return ActivityUtil(this)
        }

        init {
            initBuilder(aClazz)
        }
    }

    private fun show() {
        val lIntent = Intent(mContextActivity, mClazz)
        lIntent.putExtras(mBundle!!)
        if (mOnActivityResult) if (mContextActivity is Activity) mContextActivity.startActivityForResult(
            lIntent,
            mRequestCode
        ) else mContextActivity.startActivity(lIntent) else mContextActivity.startActivity(lIntent)
    }

    init {
        mContextActivity = aBuilder.mContextActivity
        mBundle = aBuilder.mBundle
        mClazz = aBuilder.mClazz
        mRequestCode = aBuilder.mRequestCode
        mOnActivityResult = aBuilder.mOnActivityResult
        show()
    }
}